### ISSUE TITLE
psalter incorrectly removes variable assignment in loop

### ISSUE THREAD

**Issue by ADmad** [vimeo/psalm#2355]
For code similar to one showed below, using `psalm --alter --issues=UnusedVariable` incorrectly removes the `$first = false;` statement inside the loop.

```
$first = true;
foreach ($array as $item) {
    if ($first) {
       //do something
    }
    $first = false;
}
```

**Comment by muglug** [comment]
Can you provide a link on psalm.dev to reproduce? At a minimum I believe Psalm must first wrongly identify that the variable goes unused (and it’s [not found on the code above](https://psalm.dev/r/b359bdc1b7)).

**Comment by ADmad** [comment]
That's the thing, pslam doesn't report them as unused variable, but pslater removes them. 

Here's an example from actual code https://github.com/cakephp/cakephp/blob/2b3f9261edb78c4a948706ffb65e4ece1b9a2a60/src/Http/ResponseEmitter.php#L200-L207

If you run pslater on the whole `src` folder it removes 8 such assignments. Please also check the psalm config and baseline files. To the best of my knowledge they don't contain anything which would prevent psalm from first reporting them as unused variables.

**Comment by muglug** [comment]
Sorry about that!

This could have been caught by Psalm if https://github.com/vimeo/psalm/issues/2230 were implemented and `StatementsAnalyzer::$unused_var_locations` was documented to be an associative array.

**Comment by ADmad** [comment]
I don't have enough knowledge of the internals to understand how this issue related you the referenced issue :blush:.

Thank you for your awesome response time to the issue (as always) and this wonder tool in general.

**Comment by muglug** [comment]
> I don't have enough knowledge of the internals to understand how this issue related you the referenced issue

SO

Psalm currently supports three different ways of annotating arrays:

- the generic `array<key, value>`
  - which has a subtype `list<value>` for arrays where all items are indexed by sequential integers starting at zero
- the object-like array type `array{foo: string, bar: int}`

Calling `array_reverse` [without that second argument]( e76beff) with both `list<Foo>` and `array<int, Foo>` returns a `list<Foo>`.

That PR proposes adding another subtype to `array<key, value>` of `associative-array<key, value>` where Psalm will preserve those key-value associations. Importantly, a property of type `list` could not be assigned an `associative-array`-typed value and vice-versa.

So this code would not pass Psalm's checks:

```php
class A {
  /** @var associative-array<int, Foo> */
  public array $map;

  public function reverseIt() : void {
    $this->map = array_reverse($this->map);
  }
}
```
, failing with an error like `Cannot set $this->map of type associative-array<int, Foo> to type list<Foo>`

**Comment by ADmad** [comment]
Thanks for the explanation. I understand now that if #2230 was implemented then psalm would have could the error in it's own codebase and reported the erroneous assignment in the statement you fixed in e76beffa784c9c622fd14378c91068d75da1218b :slightly_smiling_face:.
